src/ngx_http_lua_worker_thread.c

```diff
@@ -0,0 +1,497 @@
+
+    typ = lua_type(from, idx);
+    switch (typ) {
+    case LUA_TBOOLEAN:
+        lua_pushboolean(to, lua_toboolean(from, idx));
+        return LUA_TBOOLEAN;
+
+    case LUA_TLIGHTUSERDATA:
+        lua_pushlightuserdata(to, lua_touserdata(from, idx));
+        return LUA_TLIGHTUSERDATA;
+
+    case LUA_TNUMBER:
+        lua_pushnumber(to, lua_tonumber(from, idx));
+        return LUA_TNUMBER;
+
+    case LUA_TSTRING:
+        str = lua_tolstring(from, idx, &len);
+        lua_pushlstring(to, str, len);
+        return LUA_TSTRING;
+
+    case LUA_TTABLE:
+        top_from = lua_gettop(from);
+        top_to = lua_gettop(to);
+
+        lua_newtable(to);
+
+        /* to positive number */
+        if (idx < 0) {
+            idx = lua_gettop(from) + idx + 1;
+        }
+
+        lua_pushnil(from);
+
+        while (lua_next(from, idx) != 0) {
+            if (ngx_http_lua_xcopy(from, to, -2, 0) != LUA_TNONE) {
+                if (ngx_http_lua_xcopy(from, to, -1, 0) != LUA_TNONE) {
+                    lua_rawset(to, -3);
+
+                } else {
+                    lua_settop(from, top_from);
+                    lua_settop(to, top_to);
+                    return LUA_TNONE;
+                }
+
+            } else {
+                lua_settop(from, top_from);
+                lua_settop(to, top_to);
+                return LUA_TNONE;
+            }
+
+            lua_pop(from, 1);
+        }
+
+        return LUA_TTABLE;
+
+    case LUA_TNIL:
+        if (allow_nil) {
+            lua_pushnil(to);
+            return LUA_TNIL;
+        }
+        /* fall through */
+
+    /*
+     * ignore unsupported values:
+     * LUA_TNONE
+     * LUA_TFUNCTION
+     * LUA_TUSERDATA
+     * LUA_TTHREAD
+     */
+    default:
+        return LUA_TNONE;
+    }
+}
+
+
+/* executed in a separate thread */
+static void
+ngx_http_lua_worker_thread_handler(void *data, ngx_log_t *log)
+{
+    ngx_http_lua_worker_thread_ctx_t     *ctx = data;
+    lua_State                            *vm = ctx->ctx->vm;
+
+    ctx->rc = lua_pcall(vm, ctx->n_args, LUA_MULTRET, 0);
+}
+
+
+/* executed in nginx event loop */
+static void
+ngx_http_lua_worker_thread_event_handler(ngx_event_t *ev)
+{
+    ngx_http_lua_worker_thread_ctx_t *worker_thread_ctx;
+    lua_State                        *L;
+    ngx_http_request_t               *r;
+    ngx_connection_t                 *c;
+    int                               nresults;
+    size_t                            len;
+    const char                       *str;
+    int                               i;
+    int                               rc;
+    ngx_http_lua_ctx_t               *ctx;
+    lua_State                        *vm;
+    int                               saved_top;
+
+    worker_thread_ctx = ev->data;
+
+    if (worker_thread_ctx->is_abort) {
+        ngx_http_lua_free_task_ctx(worker_thread_ctx->ctx);
+        ngx_http_lua_thread_task_free(worker_thread_ctx);
+        return;
```

Comment:
we'd better use `goto` to avoid such duplicated lines, they also appear in the following code.